xarray/core/dataset.py

```diff
@@ -5246,7 +5246,13 @@ def to_stacked_array(
         """
         from xarray.structure.concat import concat
 
-        stacking_dims = tuple(dim for dim in self.dims if dim not in sample_dims)
+        # add stacking dims by order of appearance
+        stacking_dims_list: list[Hashable] = []
+        for da in self.data_vars.values():
```

Comment:
Sorry for the late review @aFarchi . We will need to iterate over `self.coords` too. We can have `coords` variables with dimensions not present on any `data_var`